backend/special_effect.py

```diff
@@ -0,0 +1,354 @@
+            arg (Object): The object that the effect is applied to. If the
+                special effect is not applied to an object, arg is None.
+        """
+        super().__init__(param, effects, completed, arg)
+        self.condition = conditions
+
+    def __eq__(self, other):
+        """
+        Checks if two conditional effects are equal.
+
+        Args:
+            other (ConditionalEffect): The conditional effect to compare to.
+
+        Returns:
+            bool: True if the effects are equal, False otherwise.
+        """
+        return self.param == other.param and self.effects == other.effects \
+            and self.completed == other.completed \
+                and self.condition == other.condition \
+                    and self.arg == other.arg
+    
+    def __hash__(self):
+        """
+        Returns the hash of the conditional effect.
+
+        Returns:
+            hash (int): The hash of the conditional effect.
+        """
+        return hash((self.param, tuple(self.effects), self.completed, 
+                     tuple(self.condition), self.arg))
+    
+    def __repr__(self):
+        """
+        Returns the string representation of the conditional effect.
+
+        Returns:
+            string (str): The string representation of the conditional effect.
+        """
+        return f"ConditionalEffect({self.param}, {self.completed}, {self.arg})"
+    
+    def apply_sfx_on_arg(self, arg, param_arg_dict):
+        """
+        Updates the conditional effect with an arg.
+
+        Args:
+            arg (Object): The argument that the conditional effect is applied to.
+            param_arg_dict (Dictionary[Object, Object]): The dictionary mapping
+                the parameters to the arguments.
+
+        Returns:
+            copy (ConditionalEffect): The updated conditional effect with an arg. 
+        """
+        new_effects = {}
+        for effect, value in self.effects.items():
+            new_effects[effect.replace_pred_params_with_args(param_arg_dict)] = value
+        new_conditions = {}
+        for condition, value in self.condition.items():
+            new_conditions[condition.replace_pred_params_with_args(param_arg_dict)] = value
+        return ConditionalEffect(self.param, new_effects, self.completed, new_conditions, arg)
+    
+    def update(self, state, active=False):
+        """
+        Updates the state with the effect.
+
+        Args:
+            state (State): The state to update.
+            active (bool): Whether or not the update is due to an action being
+            performed.
+        """
+        if active: return
+        for condition, value in self.condition.items():
+            if state.predicates[condition] != value:
+                return
+        for effect, value in self.effects.items():
+            state.update_predicate(effect, value)
+        self.completed = True
+
+class CreationEffect(SpecialEffect):
```

Comment:
when we implement this in the future let's also make separate files for each special effect. file size will get enormous if we don't 😃

Reply:
I made new files for each subclass and placed the subclasses in a folder backend.special_effects